File: 7-kyu/remove-duplicate-words/c++/solution.cpp
#include <set>
#include <string>

std::string removeDuplicateWords(const std::string& str)
{
    if (str.empty()) return "";
    else if (str.find(" ") == std::string::npos) return str;
    std::string::size_type start = 0, end;
    std::set<std::string> str_set;
    std::string ret = "", temp;
    while ((end = str.find(" ", start)) != std::string::npos) {
        temp = str.substr(start, end - start);
        if (str_set.count(temp) == 0) {
            str_set.insert(temp);
            ret += temp + " ";
        }
        start = end + 1;
    }
    temp = str.substr(start);
    if (str_set.count(temp) == 0) ret += temp;
    else ret.pop_back();
    return ret;
}

File: 7-kyu/remove-duplicate-words/javascript/solution.js
const removeDuplicateWords = (s) => {
  if (!s) {
    return "";
  }
  let ret = [];
  for (const word of s.split(' ')) {
    if (!ret.includes(word)) {
      ret.push(word);
    }
  }
  return ret.join(' ');
};